plugins/poll/lib/irv.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+class DiscoursePoll::Irv
+  def self.irv_outcome(poll_id)
+    poll = Poll.find(poll_id)
+    options = PollOption.where(poll_id: poll_id).map { |hash| { id: hash.digest, html: hash.html } }
+
+    ballot = []
+    PollVote
+      .where(poll_id: poll_id)
+      .pluck(:user_id)
+      .uniq
+      .each do |user_id|
+        ballot_paper = []
+        PollVote
+          .where(poll_id: poll_id, user_id: user_id)
+          .order(rank: :asc)
+          .each { |vote| ballot_paper << vote.poll_option.digest if vote.rank > 0 }
+        ballot << ballot_paper
+      end
```

Comment:
I doubt this doesn't end up being a serious performance concern

Reply:
I was trying to keep the current architecture, such that the ballot is executed whenever you vote so that the user can see the result immediately upon transition.

Could consider running this as a background job somehow and just showing latest results to user, but I'm wondering if that will cause more confusion? "Where is my vote?"

It would be a shame to lose the real time aspect of "vote and see"?

Reply:
OK I've revisited this code and believe I've improved runtime performance by about 2.5x:

https://github.com/discourse/discourse/pull/27155/commits/261cec043c88b99ab919092655e7649155b8fa9d

@jjaffeux @gschlager